# Rarefaction

* anvio: Run Anvio docker containers in parallel. Currently used to map short reads to gene lists. Generates series of fasta files, one per gene.

* biopython: Read in fasta files generated from anvio step from a fastas directory and generate one .tsv file with the read lengths and GC content associated with each gene / replicant, along with its contig and station.

* mysql: Runs a mysql server container and generates a database designed for performing rarefaction on the data arranged in the biopython step.

* populate: Reads the .tsv generated by the biopython step and inserts data into the database. The .tsv can be split into separate parts and run in parallel.

* rarefy: Read the data from the database and generate a series of matrices, one per ecotype, showing the coverage for each gene at each station. Each gene-station's coverage is an aggregate of a random subsample of gene reads, defined by a particular depth. For each station in each ecotype, if the total number of station's reads is less than a defined threshold, it is skipped.
